src/cpp/shared/campaign/campaign.cpp

```diff
@@ -228,6 +175,162 @@ void ExpCampaign::InitNewCampaign(int var) {
+              if(this->campaignID != 3 || this->currentMapID != 2 || hro->heroID == 67)
+                break;
+            } else {
+              if(hro->heroID == 65)
+                break;
+            }
+          }
+          hro->SetSS(campChoice->field, campChoice->amount);
+        }
+        break;
+      case 4u:
+        if(player0->numHeroes > 0)
+          hero0->army.Add(campChoice->field, campChoice->amount, -1);
+        break;
+      case 5u:
+        player0->puzzlePieces = LOBYTE(campChoice->field);
+        break;
+      case 6u: {
+        int tempInNewGameSetup = gbInNewGameSetup;
+        gbInNewGameSetup = 1;
+        if(player0->numHeroes > 0) {
+          hero0->experience += campChoice->field;
+          hero0->CheckLevel();
+        }
+        gbInNewGameSetup = tempInNewGameSetup;
+        break;
+      }
+      case 9u:
+        if(player0->numHeroes > 0) {
+          hero *hro;
+          for(int i = 0; ; ++i) {
+            if(player0->numHeroes > i) {
+              hro = &gpGame->heroes[player0->heroesOwned[i]];
+              if(this->campaignID == 3) {
+                if(this->currentMapID == 2 && gpGame->heroes[player0->heroesOwned[i]].heroID != 67)
+                  continue;
+              }
+            }
+            break;
+          }
+          hro->primarySkills[campChoice->field] += campChoice->amount;
+        }
+        break;    
+      case 10u:
+        if(player0->numHeroes > 0)
+          GiveArtifact(hero0, ARTIFACT_SPELL_SCROLL, 0, campChoice->field);
+        break;
+      case 7u:
+        break;
+    }
+  }
+  for(int award = 0; award < ELEMENTS_IN(this->awards); ++award) {
+    if(this->awards[award]) {
+      switch(award) {
+        case 1:
+          if(player0->numHeroes > 0)
+            GiveArtifact(hero0, ARTIFACT_BREASTPLATE_OF_ANDURAN, 0, -1);
+          break;
+        case 3:
+          if(player0->numHeroes > 0)
+            GiveArtifact(hero0, ARTIFACT_HELMET_OF_ANDURAN, 0, -1);
+          break;
+        case 4:
+          for(int i = 0; i < MAX_HEROES; ++i) {
+            if(gpGame->heroes[i].heroID == 61)
+              gpGame->heroes[i].Deallocate(0);
+          }
+          break;
+        case 5:
+          if(player0->numHeroes > 0)
+            GiveArtifact(hero0, ARTIFACT_BATTLE_GARB_OF_ANDURAN, 0, -1);
+          break;
+        case 8:
+          if(player0->numHeroes > 0)
+            GiveArtifact(hero0, ARTIFACT_LEGENDARY_SCEPTER, 0, -1);
+          break;
+        case 9:
+          if(player0->numHeroes > 0)
+            hero0->spellsLearned[SPELL_SET_EARTH_GUARDIAN] = 1;
+          break;
+        case 10:
+          if(player0->numHeroes > 0)
+            GiveArtifact(hero0, ARTIFACT_SPHERE_OF_NEGATION, 0, -1);
+          break;
+        case 0:
+        case 2:
+        case 6:
+        case 7:
+          continue;
+      }
+    }
+  }
+  gbRetreatWin = 1;
+}
+
+void ExpCampaign::Autosave() {
```

Comment:
What was changed in this?

Reply:
string parameter is an item from std::map